### PR TITLE
Multiuser Task List Locking Subscription

### PR DESCRIPTION
**Pull request by OddTomBrooks**
# EASI-1995

## Changes and Description

- RefCounted implementation of simple task list locking.
- Note: Nothing in this PR actually secures the transactions, that will be done as a separate pass for simplicity.

### Create a New Websocket Request
image

### Create a GQL_CONNECTION_INIT Request
image

```
{
    "type": "connection_init"
}
```

### Create an onTaskListSectionsChanged Request
image

```
{
    "payload": {
        "operationName": "onTaskListSectionLocksChanged",
        "query": "subscription onTaskListSectionLocksChanged($modelPlanID: UUID!) { onTaskListSectionLocksChanged(modelPlanID: $modelPlanID) { changeType lockStatus { modelPlanID section lockedBy refCount } } }",
        "variables": {
            "modelPlanID": "f11eb129-2c80-4080-9440-439cbe1a286f"
        }
    },
    "type": "start"
}
```

#### Query
```
mutation lockTaskListSection($modelPlanID: UUID!, $section: TaskListSection!)
{
    lockTaskListSection(modelPlanID: $modelPlanID, section: $section)
}
```

#### GraphQL Variables
```
{
  "modelPlanID": "f11eb129-2c80-4080-9440-439cbe1a286f",
  "section": "MODEL_BASICS"
}
```

#### Query
```
mutation unlockTaskListSection($modelPlanID: UUID!, $section: TaskListSection!)
{
    unlockTaskListSection(modelPlanID: $modelPlanID, section: $section)
}
```

#### GraphQL Variables
```
{
  "modelPlanID": "f11eb129-2c80-4080-9440-439cbe1a286f",
  "section": "MODEL_BASICS"
}
```

### Testing Steps
- Run the GQL_CONNECTION_INIT request to prime your connection
  - Your response should look like this ('ka' type messages are KeepAlive messages, you will receive these at regular intervals) 
image
- Run the onTaskListSectionLocksChanged request to subscribe to task list section lock changes
  - Your response should look like this
image
- Lock a task list section
  - Your response in the WebSocket tab should look like this
image
- Unlock a task list section
  - Your response in the WebSocket tab should look like this
image
- Repeat locking and unlocking until you are satisfied
  - Bear in mind that with refcounting, if you lock a task list section twice you MUST unlock it twice to release the lock. This is by design, and is guarded by WebSocket disconnects where not explicit in order to support multiple tabs.

## PR Author Review Checklist

- [ ] Met the ticket's acceptance criteria, or will meet them in a subsequent PR.
- [ ] Added or updated tests for backend resolvers or other functions as needed.
- [ ] Added or updated client tests for new components, parent components, functions, or e2e tests as necessary.
- [ ] Tested user-facing changes with voice-over and the [rotor menu](https://support.apple.com/guide/voiceover/with-the-voiceover-rotor-mchlp2719/mac)

## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.

### CODE REVIEW

**Review by ClayBenson94 — COMMENTED** [review]
Aside from the minor comment in the block above - the only thing that I think could potentially be changed here (that isn't auth related) is the following test case:

- Lock a task list section for model plan `abcd`
- Subscribe to notifications for locks on model plan `abcd`
- Remove the lock on the task list section for model plan `abcd`

From the subscriber's point of view, all they'll see is "lock removed", but they were never aware of there being a lock in the first place.

As I am writing this, I've found that `query taskListSectionLocks` exists - is the intent for the client to query this when making a new connection to help ensure that previously made locks are made aware to the client? If so, I think that works for now, until we decide to beef up the server-side validation that's being done with these locks.

**Comment by StevenWadeOddball** [comment]
In case anyone has issues testing, I had to set the authorization header again for the websocket
`Authorization `      `Local {"EUAID":"MINT","jobCodes":["MINT_D_ADMIN"],"favorLocalAuth":true}`

**Review by StevenWadeOddball — CHANGES_REQUESTED** [review]
Thanks @OddTomBrooks, overall looks good, my main thought is that perhaps we should have a model plan id in the `TaskListSectionLockStatus` type.

Otherwise, just a minor observation about the ADR name, and needing an index file update.

Thanks for the testing instructions! I tested the change by making a connection, then initiating a subscription. I was able to see the lock and unlock events. 

I did notice that you don't see an event if the lock refCount increases, but doesn't make a new lock (I assume that was intentional?). That tricked me for a few minutes, but I think it makes sense.

**Comment by ClayBenson94** [comment]
> I did notice that you don't see an event if the lock refCount increases, but doesn't make a new lock (I assume that was intentional?). That tricked me for a few minutes, but I think it makes sense.

@StevenWadeOddball I saw the same, but with refCount decreases (and not removing a lock). I think this makes sense from a subscription standpoint - the only reason that a subscriber would want to get an update is if the lock is removed or added, so I think it's alright that it waits until all refs are deleted / a new lock is created.

I think we'll probably, in the long run, be able to formalize these subscriptions to hold the client's hand a bit more, but I think this implementation does allow it to work from the FE!

File: pkg/graph/schema.graphql [hunk 1]
```diff
@@ -107,6 +107,17 @@ type ExistingModel {
...
+type TaskListSectionLockStatusChanged {
+  changeType: ChangeType!
+  lockStatus: TaskListSectionLockStatus!
+}
+
+type TaskListSectionLockStatus {
+  section: TaskListSection!
+  lockedBy: String!
+  refCount: Int!
+}
```
[the remark this answers is not in the document]
Comment (ClayBenson94) [reply pkg/graph/schema.graphql:119]: That's probably a good callout. If there's not a way to see which model plan the lock is created for, then a user with multiple model plans open in tabs would probably end up over-locking in the UI.

**Comment by StevenWadeOddball** [comment]
> 
@ClayBenson94 I completely I agree. I just more or less wanted to note the current state. I think it would probably be more confusing to get a notification for every event instead of new locks, and a section getting completely unlocked.

File: pkg/graph/schema.graphql (hunk 1 above)
[the remark this answers is not in the document]
Comment (OddTomBrooks) [reply pkg/graph/schema.graphql:119]: The general way to manage asynchronous commands like this is that it's the client's responsibility to assign appropriate callback streams per subscription. Ideally we should transfer minimal information but our minimum block size is such that we can just add it for client convenience with no real cost.

**Review by ClayBenson94 — DISMISSED** [review]

**Review by ClayBenson94 — APPROVED** [review]
Tested the new `modelPlanID` changes in the `lockStatus` locally and it looks good. Thanks @OddTomBrooks!

**Pull request merged**